### DISCUSSION TITLE
Open issues

### DISCUSSION THREAD

**Discussion by aclist** [aclist/kbin-megamod#5] [Archived]
Open issues/features anyone can feel free to pick up and elaborate

1. Gracefully handle promise response when fetching JSON metadata and prepare an error message in lieu of the array if remote is unreachable. If we have no manifest results to show, print a warning inside the modal and prompt to reload.

2. Use version number in top right as link back to project. Compare version number at page load and convert version number to a warning to update if version is out of synch. (I.e. replace innerText with a string)

3. Pagination/UI scheme mockups, tests

4. Tooltips (could be used to remove verbose description messages and gloss non-obvious labels with a description triggered onHover. Potentially opens up a lot of room, but may be less intuitive than inline descriptions. Might be worth trialing

**Comment by artillect** [comment]
I'll take a crack at 2, 3, and 4. Those last two should be pretty easy to do with jquery-ui, and I'll see what I can do about the version number checking. 

Maybe a page on github pages with the current version number would work? I don't want to load the whole script every time we do a version check. For now, I'll just add a link back here

**Reply by aclist** [reply]
Cool, thank you! 

> Maybe a page on github pages with the current version number would work? I don't want to load the whole script every time we do a version check. For now, I'll just add a link back here

Can just store a flat file in the repo and bump the version number in it.

> I have this boilerplate logic kicking around if you want to repurpose it. The `onload` method is async

```

const versionFile = "VERSION_FILE_URL";
const upstream = "REPO_URL";
const version = "VERSION";
const element = "VERSION_TIP";
element.setAttribute('href',upstream);

function checkVersion() {
    GM_xmlhttpRequest({
        method: 'GET',
        url: versionFile,
        onload: compare,
        headers: {
            "User-Agent": "Mozilla/5.0",
            "Accept": "text/xml"
        },

    });
};
function compare(response) {
    let parser = new DOMParser();
    let doc = parser.parseFromString(response.responseText, "text/html");
    let content = response.responseText
    if (content != version) {
            element.innerText "New version available!";
            element.style.cssText += 'color: pink';
            } else {
                return
            }
        }
    }
```

**Reply by aclist** [reply]
The other nice thing about a version file is that it acts as a deploy action: updating the script is decoupled from the manual version bump action, so the version number changing in the master script does not deploy to people's clients until the version file is explicitly bumped.

**Reply by artillect** [reply]
I've been working on the tabs and tooltips, and here's what I've got so far.
![image](https://github.com/aclist/kbin-megamod/assets/8906201/e333fb1b-e6fd-4b70-b8c0-0aec1f368b16)

Changes:
- Tabs
  - Moved to the side
  - Now switch between pages of options
- Option descriptions
  - Hidden by default, shown on hover
  - Moved author name to right and added a link to it (should this be to kbin or github?)

What do you think? It should be pretty easy to add static pages (like a home page/credits page) to the options with the way I've done the pagination. It just hides all of the elements in the modal's body that don't have the class corresponding to the page.

**Reply by aclist** [reply]
I like the implementation, that is basically what I was going to do with the tabs (blanking the page). The other thing I think we can eventually do here is move the section labels into the manifest and set up the entire menu programmatically so that no sidebar/page names are hardcoded, being sourced from the manifest itself. (UI manifest could be its own file if necessary).

On the visual front, I think there is too much dead space here at the moment (not sure if done to accommodate responsive layout?) and end-users are gonna tar and feather us if this falls into the uncanny valley of "a programmer made this," so I want to get things pretty well organized in the view. I looked at it at 1440p and 1080p 16:9, and at 1080p vertical (9:16).

I've made some tweaks on your PR and will add those commits into the remote WIP after I fix some bugs I introduced :D! Just updating this with a view into the changes so we don't diverge too much on the UI front before I send some commits your way.

Image and summary of changes:

![modal](https://github.com/aclist/kbin-megamod/assets/92275929/ca6c7e60-c319-4289-987c-d03ed55d8d8b)

- Breadcrumbs
   - Create breadcrumbs and move into header. I think this gives a clearer indication of where we are relative to the main settings. This can also accommodate deeper nesting if we go another layer deep, like `Megamod Settings :: Threads :: Subsection`
   - Apropos of the above: some mods or options might necessitate a whole page of their own, so should think about what that logic would look like.
    - Now that the breadcrumbs occupy the top left and the header is more visually anchored on both sides, move the docking button near the other button
- Padding    
  - Attempt to tighten padding in grid. We'll probably have to keep playing with this to get the right feel and make sure it looks reasonably acceptable on a vertical or small screen.
- Tooltips
  - Separate tooltips from checkbox inputs and make tooltips clickable (onclick event). This should prevent noisy description text from updating every time the mouse is moving around the list, and the icon should clarify that this is a help callout. The description does not change until clicked again, so it's easier to navigate to the author link without having to be pixel perfect on the line
- Detailed description
  - Created something called the `.megamod-settings-modal-helpbox` that contains the author link and description
   - This should keep the details in a more uniform size and shape and unified location and prevent the span tags in the label from causing weird alignment issues on small screens.
   - This box can also be used for more verbose descriptions or even instructions/tutorial of how to use the mod options, or explanation of what parameters do. Also gives more wiggle room for things like multiple authors, lists, etc.
   
I think this strikes a balance of verbosity and concision, so triptych view: sidebar/options/helpbox. Maybe they could be equidistant from each other, I don't know. I feel like the dead space effect gets exaggerated on larger displays. Three columns seems to be about the max we can fit on a smaller display.

As for the author link: I don't know, could link to anything, really. Could be a link back to the author's script page, github, kbin. Insofar as these are kbin users, it seems reasonable to point to their profile page. I don't have any problem with it. If the author prefers some arbitrary link, can just add that verbatim in the manifest.

I noticed the close ('x') icon wasn't flush with the HR in your image. Is that just something going on with your browser?

**Reply by aclist** [reply]
Other thing I was thinking that might keep the view from getting chaotic: helpbox can be used to store complex form fields. So the checkbox on the options body just keeps the list items aligned and is a global "enable this on/off", then the actual parameters go in the helpbox and you can granularly choose them there. 

For instance, turning on syntax highlighting is just a global on/off switch and reverts to the default, but then the helpbox would contain a dropdown that lets you choose from a range of CSS themes. Language filter is a global on/off, but the helpbox contains a field where you can explicitly choose which languages. Etc.

**Reply by aclist** [reply]
OK, UI changes are commited into your fork

**Reply by artillect** [reply]
Looks great, I think that adding a bit more space and maybe a border between the sidebar and options would help a bit. Right now it sorta looks like the checkboxes are for the pages the way they're lined up.

I think putting complex form fields in the helpbox could work, but I'm a little worried that they'd be too hidden as is. Maybe adding a chevron to indicate that more options are available in the helpbox would work? Or maybe a dropdown. Originally I thought that just leaving them in the options menu below the main toggle, with a bit of indentation, would work, but you're right that it might clutter the UI a bit too much.

**Reply by aclist** [reply]
Yes, I agree. I tried using a 1px border around each section to create a visible grid structure but didn't end up liking it for some reason, looked like a spreadsheet.

Maybe replace the question mark with a gear icon to indicate that detailed settings can be exposed from there? I figure there is no rhyme or reason to the types of input fields a mod might need, so if it's inline it becomes a collection of random widgets of different sizes.

I think it'll start to make sense once we merge the latest patch and have this framework going, then we can start adding fancier forms for each mod into the helpbox and ensuring they work.

**Reply by artillect** [reply]
> Yes, I agree. I tried using a 1px border around each section to create a visible grid structure but didn't end up liking it for some reason, looked like a spreadsheet.

I'll mess around with it some more until I come up with something that feels right. I think I'm gonna make some adjustments to the header too.

> Maybe replace the question mark with a gear icon to indicate that detailed settings can be exposed from there? I figure there is no rhyme or reason to the types of input fields a mod might need, so if it's inline it becomes a collection of random widgets of different sizes.

I think that could work, I'll swap out the icon for a gear. I don't think the different options would look too random, but the input would probably need to be moved to the right so the text is all aligned.

> I think it'll start to make sense once we merge the latest patch and have this framework going, then we can start adding fancier forms for each mod into the helpbox and ensuring they work.

I can get behind that. I think we can also start recruiting some other script authors and get their input on the UI and everything else (and hopefully get some more contributions)

**Reply by aclist** [reply]
Now that we have a working foundation, we need to adopt more atomic issue tickets and use the discussion forum for ideation. I'm going to close this thread and move the ticket mentioned in number 1 to its own stub.  Regarding the UI, everything should be more or less ready to go in the new branch, just need to finish hooking the toggle button the old logic and functionality will be the same.
